File: Problem12/boundarytrav.rb
$traversal=[]
def left_boundary(root)
    if(root==nil or (root.left==nil and root.right==nil))
        return 
    end
    $traversal.append(root.val)
    if(root.left!=nil)
        left_boundary(root.left)
    else
        left_boundary(root.right)
    end
end
def leaf_boundary(root)
    if(root==nil)
         return 
    end 
    if(root.left==nil and root.right==nil)
        $traversal.append(root.val)
        return
    end
    leaf_boundary(root.left)
    leaf_boundary(root.right)
end
def right_boundary(root)
    if(root==nil or (root.left==nil and root.right==nil))
        return 
    end
    if(root.right!=nil)
        right_boundary(root.right)
    else
        right_boundary(root.left)
    end
    $traversal.append(root.val)
end
def boundary_of_binary_tree(root)
    if $traversal!=nil
        $traversal.clear
    end
    if root==nil
        return $traversal
    end
    $traversal.append(root.val)
    left_boundary(root.left)
    leaf_boundary(root.left)
    leaf_boundary(root.right)
    right_boundary(root.right)
    return $traversal
end

File: Problem12/boundarytrav.py
from ast import List
from typing import Optional
class TreeNode:
     def __init__(self,val=0,left=None,right=None):
         self.val=val
         self.left=left
         self.right=right
class Solution:
    traversal=list()
    def left_boundary(root):
        if root is None or (root.left is None and root.right is None):
            return 
        Solution.traversal.append(root.val)
        if root.left is not None :
            Solution.left_boundary(root.left)
        else:
            Solution.left_boundary(root.right)
    def leaf_boundary(root):
        if root is None:
            return 
        if root.left is None and root.right is None:
            Solution.traversal.append(root.val)
            return 
        Solution.leaf_boundary(root.left)
        Solution.leaf_boundary(root.right)
    def right_boundary(root):
        if root is None or ( root.left is None and root.right is None):
            return 
        if root.right is not None:
            Solution.right_boundary(root.right)
        else:
            Solution.right_boundary(root.left)
        Solution.traversal.append(root.val)
            
    def boundaryOfBinaryTree(self, root: Optional[TreeNode]) -> List[int]:
        Solution.traversal.clear()
        if root is None:
            return Solution.traversal
        Solution.traversal.append(root.val)
        Solution.left_boundary(root.left)
        Solution.leaf_boundary(root.left)
        Solution.leaf_boundary(root.right)
        Solution.right_boundary(root.right)
        return Solution.traversal